streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -378,42 +320,88 @@ public void prepareCommit() {
+            case RESTORING:
+                committableOffsets = Collections.emptyMap();
 
-                if (!eosEnabled) {
-                    stateMgr.checkpoint(checkpointableOffsets());
-                }
+                break;
 
-                log.debug("Committed");
+            case RUNNING:
+            case SUSPENDED:
+                final Map<TopicPartition, Long> partitionTimes = extractPartitionTimes();
+
+                committableOffsets = new HashMap<>(consumedOffsets.size());
+                for (final Map.Entry<TopicPartition, Long> entry : consumedOffsets.entrySet()) {
+                    final TopicPartition partition = entry.getKey();
+                    Long offset = partitionGroup.headRecordOffset(partition);
+                    if (offset == null) {
+                        try {
+                            offset = mainConsumer.position(partition);
+                        } catch (final TimeoutException error) {
+                            // the `consumer.position()` call should never block, because we know that we did process data
+                            // for the requested partition and thus the consumer should have a valid local position
+                            // that it can return immediately
+
+                            // hence, a `TimeoutException` indicates a bug and thus we rethrow it as fatal `IllegalStateException`
+                            throw new IllegalStateException(error);
+                        } catch (final KafkaException fatal) {
+                            throw new StreamsException(fatal);
+                        }
+                    }
+                    final long partitionTime = partitionTimes.get(partition);
+                    committableOffsets.put(partition, new OffsetAndMetadata(offset, encodeTimestamp(partitionTime)));
+                }
 
                 break;
 
+            case CLOSED:
+                throw new IllegalStateException("Illegal state " + state() + " while getting commitable offsets for active task " + id);
+
+            default:
+                throw new IllegalStateException("Unknown state " + state() + " while post committing active task " + id);
+        }
+
+        return committableOffsets;
+    }
+
+    @Override
+    public void postCommit() {
+        switch (state()) {
             case RESTORING:
-                commitNeeded = false;
+            case RUNNING:
+            case SUSPENDED:
                 commitRequested = false;
+                commitNeeded = false;
+
+                if (state() == State.RESTORING || state() == State.SUSPENDED) {
+                    writeCheckpointIfNeed();
+                } else if (!eosEnabled) { // if RUNNING, checkpoint only for non-eos
+                    writeCheckpointIfNeed();
+                }
 
-                stateMgr.checkpoint(checkpointableOffsets());
+                if (state() == State.SUSPENDED) {
+                    partitionGroup.clear();
```

Comment:
Hmm... this reads a bit weird to me. Can we call this in `suspend` instead? Also in that case we do not need to call this in close and closeAndRecycle.

Reply:
We cannot call it in suspend, because we would loose the partition-time information that we need in `prepareCommit()` (that is called after `suspend()`).